src/app/InteractionModelEngine.cpp

```diff
@@ -485,6 +540,217 @@ void InteractionModelEngine::AddReadClient(ReadClient * apReadClient)
+         usedAttributePaths + aRequestedAttributePathCount > attributePathCap) ||
+        (aRequestedEventPathCount > static_cast<int32_t>(kMinSupportedPathPerSubscription) &&
+         usedEventPaths + aRequestedEventPathCount > eventPathCap))
+    {
+        // We cannot offer enough resources, and the subscription is requesting more than the spec limit.
+        return false;
+    }
+
+    bool didEvictHandler = true;
+
+    const auto evictAndUpdateResourceUsage = [&](FabricIndex fabricIndex, bool forceEvict) {
+        bool ret           = TrimFabric(fabricIndex, forceEvict);
+        usedAttributePaths = static_cast<int32_t>(mAttributePathPool.Allocated());
+        usedEventPaths     = static_cast<int32_t>(mEventPathPool.Allocated());
+        usedReadHandlers   = static_cast<int32_t>(mReadHandlers.Allocated());
+        return ret;
+    };
+
+    //
+    // At this point, we have an inbound request that respects minimas but we still don't have enough resources to handle it. Which
+    // means that we definitely have handlers on existing fabrics that are over limits and need to evict at least one of them to
+    // make space.
+    //
+    while ((usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+            usedEventPaths + aRequestedEventPathCount > eventPathCap || usedReadHandlers >= readHandlerCap) &&
+           didEvictHandler)
+    {
+        // There might be cases that one fabric has lots of subscriptions with one interested path, while the other fabrics are not
+        // using excess resources. So we need to do this multiple times until we have enough room or no fabrics are using excess
+        // resources.
+        didEvictHandler = false;
+        for (const auto & fabric : *mpFabricTable)
+        {
+            didEvictHandler = didEvictHandler || evictAndUpdateResourceUsage(fabric.GetFabricIndex(), false);
+            if (usedAttributePaths + aRequestedAttributePathCount <= attributePathCap &&
+                usedEventPaths + aRequestedEventPathCount <= eventPathCap && usedReadHandlers < readHandlerCap)
+            {
+                break;
+            }
+        }
+    }
+
+    // We either have enough resources, or the resource usage from all fabrics are within the quota, so evict one (possibly the
+    // oldest) subscription from the current fabric.
+    didEvictHandler = true;
+    while ((usedAttributePaths + aRequestedAttributePathCount > attributePathCap ||
+            usedEventPaths + aRequestedEventPathCount > eventPathCap || usedReadHandlers >= readHandlerCap) &&
+           // Avoid infinity loop
+           didEvictHandler)
+    {
+        didEvictHandler = evictAndUpdateResourceUsage(aFabricIndex, true);
+    }
+
+    if (!didEvictHandler)
+    {
+        // This should never happen, since we should always be able to serve the request after evicting all subscriptions on the
+        // current fabric.
```

Comment:
Is this a `VerifyOrDie` scenario? I presume it is, since we have a well-behaved subscription that respected minimas, but somehow couldn't find space for it.